superchain/chain.go

```diff
@@ -0,0 +1,138 @@
+package superchain
+
+import (
+	"archive/zip"
+	"bytes"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"path"
+	"sort"
+	"sync"
+
+	"github.com/BurntSushi/toml"
+	"github.com/klauspost/compress/zstd"
+)
+
+//go:embed superchain-configs.zip
+var configData []byte
+
+var configDataReader *zip.Reader
+
+var genesisZstdDict []byte
+
+var Chains = make(map[uint64]*Chain)
+
+var idsByName = make(map[string]uint64)
+
+func ChainIDByName(name string) (uint64, error) {
+	id, ok := idsByName[name]
+	if !ok {
+		return 0, fmt.Errorf("unknown chain %q", name)
+	}
+	return id, nil
+}
+
+func ChainNames() []string {
+	var out []string
+	for _, ch := range Chains {
+		out = append(out, ch.Name+"-"+ch.Network)
+	}
+	sort.Strings(out)
+	return out
+}
+
+func GetChain(chainID uint64) (*Chain, error) {
+	chain, ok := Chains[chainID]
+	if !ok {
+		return nil, fmt.Errorf("unknown chain ID: %d", chainID)
+	}
+	return chain, nil
+}
+
+type Chain struct {
+	Name    string `json:"name"`
+	Network string `json:"network"`
+
+	config  *ChainConfig
+	genesis []byte
+
+	initOnce sync.Once
+	err      error
+}
+
+func (c *Chain) Config() (*ChainConfig, error) {
+	c.initOnce.Do(c.populate)
+	return c.config, c.err
+}
+
+func (c *Chain) GenesisData() ([]byte, error) {
+	c.initOnce.Do(c.populate)
+	return c.genesis, c.err
+}
+
+func (c *Chain) populate() {
```

Comment:
Would be nice to log the duration it took to execute `populate` at the end, so we can easier debug performance issues if they were to be caused by this.